### ISSUE TITLE
Getting System.IndexOutOfRangeException when processing results of a batch request

### ISSUE THREAD

**Issue by techfooninja** [unchase/Unchase.Odata.Connectedservice#45]
### Describe the bug
When submitting a batch request (specifically a batch of PATCH requests), the request succeeds, but the client throws a System.IndexOutOfRangeException. Repros > 90% of the time for our code. I noticed the issue went away when I moved back to Microsoft.OData.ConnectedService.

### Assemblies affected

Generated OData v4 Client
Using Microsoft.OData.Client 7.6.4

### Steps to reproduce

1. Have an OData API available - I'm currently using Microsoft.AspNet.OData 7.4.0
2. Generate OData client using Unchase (I'm using Microsoft.OData.Client 7.6.4)
3. Create a DataServiceCollection for one of the entities (i.e. `var collection = new DataServiceCollection<EntityType>(context);`)
4. Load several entities to the DataServiceCollection (i.e. using `collection.Load(entity)`) and modify them (I am modifying existing entities and adding new entities in my code)
5. Call `context.SaveChanges(SaveChangesOptions.BatchWithSingleChangeset)`

### Expected result

Batch API call succeeds and client handles response correctly.

### Actual result

Batch API call succeeds (confirmed in fiddlers), but client fails on handling the response with the following exception:

Microsoft.OData.Client.DataServiceRequestException: An error occurred while processing this request. ---> System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Array.Clear(Array array, Int32 index, Int32 length)
   at System.Collections.Generic.HashSet`1.Clear()
   at Microsoft.OData.Client.BaseSaveResult.HandleResponsePut(Descriptor descriptor, HeaderCollection responseHeaders)
   at Microsoft.OData.Client.BaseSaveResult.HandleOperationResponse(Descriptor descriptor, HeaderCollection contentHeaders)
   at Microsoft.OData.Client.BatchSaveResult.<HandleBatchResponse>d__26.MoveNext()
   --- End of inner exception stack trace ---
   at Microsoft.OData.Client.BatchSaveResult.HandleBatchResponseInternal(ODataBatchReader batchReader)
   at Microsoft.OData.Client.BatchSaveResult.HandleBatchResponse()
   at Microsoft.OData.Client.BatchSaveResult.HandleResponse()
   at Microsoft.OData.Client.BaseSaveResult.EndRequest()
   at Microsoft.OData.Client.DataServiceContext.SaveChanges(SaveChangesOptions options)
   at CALLING_CODE

### Additional detail

This only fails if I generate the client with the Unchase client. When I go back to using Microsoft.OData.ConnectedService (0.9.0) it starts succeeding again.

**Comment by unchase** [comment]
Hi, @techfooninja 

At the moment, I am focused on improving the [Connected service](https://github.com/OData/ODataConnectedService) by Microsoft and actively contribute to it, since its development has resumed. All features will be transferred to this project. In addition, it is developing faster, so all new features will be added there.

**Comment by techfooninja** [comment]
@unchase , not a problem then! If the answer right now is "use Microsoft's implementation", then that will work for me. I just switched over to this implementation a bit ago because it was so highly recommended on forums. I have no issues with using the Microsoft-supported one, and happy to see that you're contributing to it (I think I saw you already worked on an issue I reported over there 👍 )

thanks!
